### PR TITLE
Starting a recipe for ESC50

### PR DESCRIPTION
**Pull request by ycemsubakan**
Starting a recipe for ESC50. Not ready to merge yet, but wanted to start the work! 

I used the UrbanSound8k recipe as the base. 

I am seeing the ecapa-tdnn network is able to get ~50 % accuracy on test, and ~60 % accuracy on valid, when trained on folds [1, 2, 3] and fold 4 is used validation set.

### CODE REVIEW

**Comment by ycemsubakan** [comment]
Here's some of the things I think we should fix: (probably not exhaustive but let's start with this)

- We should write the docstrings for all classes and functions
- We should upload the CNN14 model to hugging face. 
- We should remove all the commented out code. 
- `train_nmf.py` code should support batchsize bigger than 1. 
-  All the `custom_models.py` stuff should be moved to lobes. 
- We should remove / implement the augmentation part of the code. 
- We should write a README, and provide a shell script to run the scripts in order. 
- We should remove all the librosa stuff. 
- All the soundfile stuff needs to be replaced with torchaudio. 
- Data saving should be made optional. 
- `--test_only` option should be implemented. 
- `--top_3_fidelity` should be named `top_k_modality` and k should included in the hparams file. 
- We should reuse the `dataio_prep` function from `train.py`.
- The `train_interpreter.py` function should be called `train_l2i.py`.

File: recipes/ESC50/classification/hparams/interpreter.yaml [hunk 1]
```diff
@@ -27,6 +27,7 @@ save_folder: !ref <output_folder>/save
 train_log: !ref <output_folder>/train_log.txt
 
 test_only: False
+save_interpretations: False
```
Comment (fpaissan) [inline recipes/ESC50/classification/hparams/interpreter.yaml:30]: defaults to false... needs to be specified in the README

**Comment by ycemsubakan** [comment]
The updated state of affairs:

- [x] We should write the proper docstrings for all classes and functions. For the classes inside lobes, we should write proper docstring tests as well..
- [ ] We should upload the pretrained CNN14 model to huggingface. 
- [ ] We should remove all the commented out code. 
- [ ] `train_nmf.py` code should support batchsize bigger than 1. 
- [x] All the `custom_models.py` stuff should be moved to lobes. 
- [x] We should remove / implement the augmentation part of the code. 
- [ ] We should write a README, and provide a shell script to run the scripts in order. 
- [x] We should remove all the librosa stuff. 
- [x] All the soundfile stuff needs to be replaced with torchaudio. 
- [x] Data saving should be made optional. 
- [x] `--test_only` option should be implemented. 
- [x] `--top_3_fidelity` should be named `top_k_modality` and k should included in the hparams file. 
- [x] We should reuse the `dataio_prep` function from `train.py`.
- [x] The `train_interpreter.py` script should be called `train_l2i.py`. 
- [x] The recipe should automatically download ESC50 if it isn't already downloaded. 
- [ ] We need to update on the logger: `top_3_fid` --> `top_k_fid`. 
- [ ] We need to clean-up the `.yaml` files so that they do include outdated comments inherited from the other recipes. 
- [ ] We need to add the trained model files to google drive, and update `recipes.csv`. 
- [ ] We need to add the authors at the top of new files. 
- [ ] We need to remove unneeded `.yaml` files.

File: speechbrain/lobes/models/L2I.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+    def forward(self, inp):
+        """This forward function returns the NMF time activations given classifier activations
+        Argument:
+            inp: A length 3 list of classifier input representions.
+        """
+        error = "in PSI doesn't match. The embedding dimensions need to be consistent with the list self.in_emb_dims"
+        for i, in_emb_dim in enumerate(self.in_emb_dims):
+            # sanity check on shapes
+            assert inp[i].shape[1] == self.in_emb_dims[i], (
```
Comment (ga1aXy-pog) [inline speechbrain/lobes/models/L2I.py:71]: assert inp[i].shape[0,1] == self.in_emb_dims[i]

**Review by ga1aXy-pog — APPROVED** [review]

**Comment by anautsch** [comment]
Please fetch & merge the latest develop version into this PR.

**Comment by TParcollet** [comment]
@ycemsubakan can you comply with the new recipe testing? We need to be able to test this recipe as well :-)

**Comment by fpaissan** [comment]
hey @TParcollet @mravanelli, just finished fixing the recipe testing. now everything is ready for review! 😃

File: recipes/ESC50/classification/confusion_matrix_fig.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * David Whipps 2021
+ * Ala Eddine Limame 2021
+"""
+
+import numpy as np
+import matplotlib.pyplot as plt
+import itertools
+
+
+def create_cm_fig(cm, display_labels):
```
Comment (anautsch) [inline recipes/ESC50/classification/confusion_matrix_fig.py:14]: Please provide docstrings with args & returns.

File: recipes/ESC50/esc50_prepare.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+    folders: list
+        Folders to check.
+    """
+    for folder in folders:
+        if not os.path.exists(folder):
+            return False
+    return True
+
+
+def full_path_to_audio_file(data_folder, slice_file_name, fold_num):
```
Comment (Adel-Moumen) [inline recipes/ESC50/esc50_prepare.py:293]: please add `data_folder` in docstring args

File: speechbrain/lobes/models/L2I.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            nn.Conv2d(out_c * 3, out_c, kernel_size=3, padding="same"),
+            nn.BatchNorm2d(out_c),
+            nn.ReLU(),
+        )
+
+        self.act = nn.ReLU()
+
+    def forward(self, inp):
+        """This forward function returns the NMF time activations given classifier activations
+        Argument:
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/L2I.py:66]: please refactor the docstring to match our i.e., 
```
Arguments
---------
```

File: speechbrain/lobes/models/PIQ.py [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+            VQ-dictionary for quantization. Expected shape of `torch.Size([K, C])` with K dictionary elements.
+        labels : torch.Tensor
+            Classification labels. Used to define irrelevant regions and divide the latent space based on predicted class. Shape should be `torch.Size([B])`.
+        num_classes : int
+            Number of possible classes
+        activate_class_partitioning : bool
+            `True` if latent space should be quantized for different classes.
+        shared_keys : int
+            Number of shared keys among classes.
+        traing : bool
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/PIQ.py:119]: typo in name `traing` -> `training`

File: speechbrain/lobes/models/PIQ.py [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+    shared_keys : int
+        Number of shared keys among classes.
+
+    """
+
+    def __init__(
+        self,
+        K,
+        D,
+        numclasses=50,
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/PIQ.py:536]: in the docstring this variable is `num_classes`, here it's `numclasses`

**Review by anautsch — APPROVED** [review]
lgtm
